### PR TITLE
Implement std::error::Error for PyErr

### LINKED ISSUE #682: PyErr does not implement std::error::Error
PyErr is a very odd structure. It pretends to emulate the 3-tuple returned by `PyErr_Fetch`, but then its `pvalue` is a 3-variant that also allows lazy construction of exception objects.

As thus it fails to implement a couple of traits typically expected of the error type (`std::error::Error`, `std::fmt::Display`, `Sync` and `Send`) and does not integrate very well with the error handling story in Python.

There are a couple of ways to go forward here, but one of the simpler things that could be done I think is:

1. Add a method along the lines of `fn(Python, PyErr) -> Py<BaseException>`;
    * I don’t believe this can be done easily right now without unsafe code.
2. Implement the typical error traits for `Py<ExceptionT>` types.
    * Specifically for `Display` and friends I think it is fine to just do a `Python::acquire_gil()` inside if that is necessary to format the data.

Willing to contribute the necessary code to implement this. Please let me know what you think about this. Perhaps other approaches seem better?

### LINKED ISSUE #1034: Question: How to convert Python Exception (PyErr) into string for logging
I have a python script which raises exception. How do I print this in the rust log?

### PR DESCRIPTION
**Pull request by davidhewitt**
This is a rework of `PyErr` to implement `std::error::Error` (and `std::fmt:Display`).

The rework was necessary because functions like `PyErr::normalize()` taking `&mut self` and `PyErr::instance()` taking `self` were incompatible with `std::error::Error`.

A couple of notes about the implementation:
- `Display` for `PyError` automatically uses `Python::with_gil` internally. I think as this is the error path, and also it's rare that the GL will not already be acquired, this overhead is fine.
- `PyErr` now contains an `UnsafeCell`. The safety is assured because, once normalized, the `UnsafeCell` value never changes. Almost every nontrivial access to the `PyErr` immediately normalizes it.

Closes #682
Closes #1034 

TODO: 
- [x] ~~Lots of documentation / guide updates!~~
- [x] ~~Add more test coverage.~~
- [x] ~~Rename `PyConcreteException::py_err` (... to `PyConcreteException::new_err` ?)~~
- [ ] Add `std::error::Error::cause()` ?

### CODE REVIEW

**Comment by kngwyu** [comment]
Thank you for this hard work and refactoring!

It's well designed overall, but I want to list up some design questions before reviewing in detail:
- `PyErrValue`. Now it's only used in `Lazy` variant, so maybe another name is suitable. In addition, maybe it's better to have `Lazy { ptype: Py<PyType>, value: Option<PyErrValue> }` instead of `PyErrValue::None`.
- `Normalizing` variant of `PyErrState`. I think some other ways (`Option` or boolean flag) also work and should be considered.
- Visibility. Though I haven't checked it precisely, are `PyErrState` and `PyErrValue` need to be `pub`?

**Comment by davidhewitt** [comment]
Thanks, good points:

- I had kept `None` / `Normalizing` in their respective enums because I thought it was more layout-efficient. But from a quick test I see that rustc is clever enough to optimize the size of `Option<SomeEnum>`:

   ```
   std::mem::size_of::<PyErrState>() = 40
   std::mem::size_of::<Option<PyErrState>>() = 40
   std::mem::size_of::<PyErrValue>() = 24
   std::mem::size_of::<Option<PyErrValue>>() = 24
   ```

   So I think in both cases let's use `Option`.

- Visibility: You're correct, I think both `PyErrValue` and `PyErrState` can be `pub(crate)`.

I'll push a commit tweaking as suggested tonight.

**Comment by davidhewitt** [comment]
(still planning to push fix later, sorry didn't get round to it yesterday.)

**Comment by davidhewitt** [comment]
This is ready for a first round of review. Once we've made any design changes necessary, I also need to write more tests and the migration guide entry before we can merge this.

File: src/exceptions.rs
Comment (birkenfeld) [inline src/exceptions.rs:614]: if `PyErr` now implements Display, shouldn't that work without the `instance` call?

**Review by birkenfeld — COMMENTED** [review]
While overhauling `PyErr`, can we fix the pretty confusing and duplicate APIs around creating new errors?

I mean mostly `PyConcreteError::py_err` and `PyConcreteError::into`:

* for me, the first doesn't really make the association with `PyErr`. I'd much rather this be called `new`: AFAICT, that method is not defined currently.

* the second reads really strange, since I'm hardwired to think of `into()` as a self-consuming method.

The two are also named completely differently, although the do very similar things (AFAICS, `into()` is just `Err(py_err())`).  IMO `into()` can be removed, I can't think of any other library that provides this shortcut.

I also think the `PyErr: Into<PyResult>` impl is unidiomatic and should be removed.

**Comment by davidhewitt** [comment]
👍 I'm completely happy to rework `PyException::py_err` and `PyException::into`. I was wondering along similar lines.

I'm definitely in favour of removing `.into` and agree on the impls being unidiomatic, so I'll remove those immediately.

Regarding `py_err` vs `new`, the only thing that makes me unsure about this is that `PyConcreteException::new()` could also return `&PyConcreteException` similar to all the rest of the `PyType::new()`. Perhaps `PyConcreteException::new_err() -> PyErr` is a less ambiguous choice?

**Comment by birkenfeld** [comment]
Yes, I was thinking about that too. `new_err` is certainly nicer than `py_err`, but since the concrete exception structs are just void structs, I don't think you can have a useful other definition of `new` there?

**Comment by davidhewitt** [comment]
Since #1024 it's possible to have a reference to all the exception objects in just the same way as for the rest of the native types (i.e. they're no longer void structs).

**Comment by davidhewitt** [comment]
The last piece of the puzzle, which I haven't finished yet but plan to do soon, is have a `#[pyexception]` macro which is pretty much just `#[pyclass]` with extra goodness to make exception types.

**Comment by birkenfeld** [comment]
Oops, I missed that. Yes, then it makes sense not to call that `new`.

**Comment by kngwyu** [comment]
`new_err` sounds good to me, though I mostly use `PyErr::new` and don't have any strong opinion about how the exception constructor would be.

**Comment by birkenfeld** [comment]
BTW, could `PyErr::new` be made nicer to use with turbofish by making `V` an `impl ...` parameter?

**Comment by davidhewitt** [comment]
Sadly it's not possible to combine turbofish with `impl Trait` parameters: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=81e2e2f20891dbb7456b3480b2fede79

File: src/err/err_value.rs
Comment (kngwyu) [inline src/err/err_value.rs:3]: How about moving this to `err_state.rs`?

File: src/err/mod.rs
Comment (kngwyu) [inline src/err/mod.rs:172]: Nice examples :+1:

**Comment by davidhewitt** [comment]
One last point of discussion: At the moment there's no implementation for `std::error::Error::cause()`, because it needs to return `Option<&dyn std::error::Error>`.

At the moment we only offer this implementation for `&PyBaseException`, because we can attach the cause object to the GIL lifetime.

I think it should be possible to implement for `PyErr`. We'd need to have a kind of "mini" object storage which the `.cause()` could be stored in, which could be cleaned up in `impl Drop for PyErr`. It feels complicated to get right, which is why I wanted to separate it from the rest of this PR (if we choose to implement it at all).

File: src/err/mod.rs
Comment (kngwyu) [inline src/err/mod.rs:424]: Nicer than the previous one with `transmute` :+1:

**Review by kngwyu — APPROVED** [review]
:100:

**Comment by kngwyu** [comment]
> I think it should be possible to implement for PyErr. We'd need to have a kind of "mini" object storage which the .cause() could be stored in, which could be cleaned up in impl Drop for PyErr.

But what is the `source`? `PyErrValue`? Or Python's backtrace?

File: src/err/mod.rs
Comment (birkenfeld) [inline src/err/mod.rs:79]: Nit: constructors -> constructor

**Comment by davidhewitt** [comment]
> But what is the source? PyErrValue? Or Python's backtrace?

It would be the normalized python exceptions's cause, as seen at https://docs.python.org/3/c-api/exceptions.html#c.PyException_GetCause

**Comment by davidhewitt** [comment]
Thanks for the comments, updated as suggested.

I'll start writing docs and tests later 👍

**Comment by davidhewitt** [comment]
Rebased, added some further test coverage, and also simplified `PyErrState` further.

I was able to take out the `PyErrValue` enum completely, and changed `PyErrArguments` to use `IntoPy` rather than `ToPyObject` (which means less copying!).

File: src/gil.rs
Comment (davidhewitt) [inline src/gil.rs:369]: `rust-numpy` will need to migrate to `Python::with_gil`

**Comment by davidhewitt** [comment]
Right, I've just pushed a load of documentation for this. I now think this is ready to merge, so would appreciate some final reviews from anyone who's got a few minutes!

File: guide/src/migration.md
Comment (kngwyu) [inline guide/src/migration.md:38]: Inserting markdown link would be better than just 'the next section'

File: guide/src/migration.md
Comment (kngwyu) [inline guide/src/migration.md:43]: Is 'not idiomatic' a too strong argument? Just 'redundant' would make sense.

File: guide/src/migration.md
Comment (kngwyu) [inline guide/src/migration.md:60]:
```suggestion
makes it possible to interact with Python exception objects.
```

File: guide/src/migration.md
Comment (kngwyu) [inline guide/src/migration.md:59]: I feel ` real definitions similar to ...` a bit difficult.
What is important is we can use them like `&PyAny`, so I think `these type now has real definitions and is usable like PyAny, PyDict etc.` would be better.

File: guide/src/migration.md
Comment (kngwyu) [inline guide/src/migration.md:27]: I think it's beneficial to concretely note what has been made private. I.e. we cannot access `err.ptype/pvalue/ptraceback` and should use `err.ptype()/pvalue()/ptraceback()`.

File: guide/src/migration.md
Comment (kngwyu) [inline guide/src/migration.md:28]: I think this is not *perfectly* true: isn't it more of a safety reason considering normalization?
And it would be maybe too detailed for just a migration guide.

**Review by kngwyu — APPROVED** [review]
LGTM, but it would be better if we have another review for the document.
@birkenfeld 
Could you take a glance on it?

**Comment by davidhewitt** [comment]
Thanks for the great feedback on the docs; I've pushed some quick changes based on your review.

**Comment by davidhewitt** [comment]
I'd like to merge this tonight; maybe @sebpuetz is interested in reviewing?

**Comment by sebpuetz** [comment]
Sure, I should find some time within the next hour

File: src/err/err_state.rs
Comment (sebpuetz) [inline src/err/err_state.rs:47]: `PyErrState` is `pub (crate)` but this fn is `pub`, given that this function returns raw pointers, I'd expect that this is not used externally. Could this also be marked `pub (crate)`? Otherwise, adding some documentation would be nice!

File: src/err/mod.rs
Comment (sebpuetz) [inline src/err/mod.rs:96]: I'm not sure whether I like this kind of assertion here, on the other hand making the error-creation fallible seems weird...

At least the requirements for `ty` should be documented.

Comment (sebpuetz) [reply src/err/mod.rs:96]: I just saw that the constructor below handles this case by returning a `PyTypeError`, perhaps this is also the right solution here?

File: src/err/mod.rs
Comment (sebpuetz) [inline src/err/mod.rs:271]: Does this take ownership of the `CString` or does `PyErr_NewException` copy the string? Else, I'm not sure how the pointer stays valid after `null_terminated_name` goes out of scope.

**Review by sebpuetz — APPROVED** [review]
I didn't have time to do a very thorough review but this looks pretty good, just a handful of questions from my side.

File: src/err/mod.rs
Comment (davidhewitt) [reply src/err/mod.rs:271]: It copies the string. (I believe this is the norm for all C APIs - it's pretty scary to steal ownership of things in C unless clearly documented, lest double-free!)

Comment (sebpuetz) [reply src/err/mod.rs:271]: Knowing this could also have some implications for CStrings in other interfaces like docstrings

File: src/err/mod.rs
Comment (davidhewitt) [reply src/err/mod.rs:96]: 👍 refactored to use the same `PyTypeError`.

**Comment by davidhewitt** [comment]
Thanks for the review! In the end I didn't have time to merge yesterday evening. I've just pushed tweaks to address your comments and will merge shortly...

**Pull request merged**
